Compression/Huffman.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * Huffman Coding is a lossless data compression algorithm that uses variable-length codes to represent characters.
+ *
+ * The algorithm works by assigning shorter codes to characters that occur more frequently. This results in a compressed representation of the data.
+ *
+ * Huffman Coding is widely used in a variety of applications, including file compression, data transmission, and image processing.
+ *
+ * More information on Huffman Coding can be found here: https://en.wikipedia.org/wiki/Huffman_coding
+ */
+
+/**
+ * Builds a frequency table from a string.
+ * @example
+ * buildFrequencyTable('this is an example for huffman encoding')
+ * returns { ' ': 6, a: 2, c: 1, d: 1, e: 4, f: 3, g: 1, h: 2, i: 3, l: 1, m: 1, n: 4, o: 1, p: 1, r: 1, s: 2, t: 2, u: 1, x: 1 }
+ * @param {string} data - The string to build the frequency table from.
+ * @returns {Object} - The frequency table.
+ */
+function buildFrequencyTable(data) {
+  const freqTable = {}
+
+  for (const char of data) {
+    freqTable[char] = (freqTable[char] || 0) + 1
+  }
+
+  return freqTable
+}
+
+/**
+ * A Huffman Node is a node in a Huffman tree.
+ * @class HuffmanNode
+ * @property {string} char - The character represented by the node.
+ * @property {number} freq - The frequency of the character.
+ * @property {HuffmanNode} left - The left child of the node.
+ * @property {HuffmanNode} right - The right child of the node.
+ */
+class HuffmanNode {
+  constructor(char, freq) {
+    this.char = char
+    this.freq = freq
+    this.left = null
+    this.right = null
+  }
+}
+
+/**
+ * Builds a Huffman tree from a frequency table.
+ * @param {Object} freqTable - The frequency table to use for building the tree.
+ * @returns {HuffmanNode} - The root node of the Huffman tree.
+ */
+function buildHuffmanTree(freqTable) {
+  const nodes = Object.keys(freqTable).map(
+    (char) => new HuffmanNode(char, freqTable[char])
+  )
+
+  while (nodes.length > 1) {
+    nodes.sort((a, b) => a.freq - b.freq)
```

Comment:
Why do you sort in every iteration?

Reply:
because with the push on a new node, the order may change.

Reply:
Wouldn't it be appropriate to use a max heap here (our existing implementations should work, you just need to import and use them), given that you always extract the most frequent ones?

Reply:
I think it would add a lot more complexity, wouldn't it?

Reply:
I think the API of our heap should be pretty straightforward to use. A heap is pretty much *the* data structure for this use case; it would significantly help the time complexity.

Reply:
I'll try to implement it myself, then I'll need your valuable feedback and guidance.